include/open62541pp/types/Variant.h

```diff
@@ -287,92 +265,80 @@ class Variant : public TypeWrapper<UA_Variant, UA_TYPES_VARIANT> {
         setScalarCopyImpl(value, dataType);
     }
 
-    /// Assign array to variant.
-    template <typename T>
-    void setArray(Span<T> array) noexcept {
-        using ValueType = typename decltype(array)::value_type;
-        assertIsNative<ValueType>();
-        setArray(array, opcua::getDataType<ValueType>());
-    }
-
-    /// @overload
-    template <typename ArrayLike, typename = EnableIfNoSpan<ArrayLike>>
+    /**
+     * Assign array to variant.
+     * @param array Container with a contiguous sequence of elements.
+     *              For example `std::array`, `std::vector` or `Span`.
+     *              The underlying array must be accessible with `std::data` and `std::size`.
+     */
+    template <typename ArrayLike>
     void setArray(ArrayLike&& array) noexcept {
-        setArray(Span{std::forward<ArrayLike>(array)});
-    }
-
-    /// Assign array to variant with custom data type.
-    template <typename T>
-    void setArray(Span<T> array, const UA_DataType& dataType) noexcept {
-        setArrayImpl(array.data(), array.size(), dataType, UA_VARIANT_DATA_NODELETE);
+        using ValueType = typename std::remove_reference_t<ArrayLike>::value_type;
+        assertIsNative<ValueType>();
+        setArray(std::forward<ArrayLike>(array), opcua::getDataType<ValueType>());
     }
 
-    /// @overload
-    template <typename ArrayLike, typename = EnableIfNoSpan<ArrayLike>>
+    /**
+     * Assign array to variant with custom data type.
+     * @copydetails setArray
+     * @param dataType Custom data type.
+     */
+    template <typename ArrayLike>
     void setArray(ArrayLike&& array, const UA_DataType& dataType) noexcept {
-        setArray(Span{std::forward<ArrayLike>(array)}, dataType);
-    }
-
-    /// Copy array to variant.
-    template <typename T>
-    void setArrayCopy(Span<T> array) {
-        using ValueType = typename decltype(array)::value_type;
-        assertIsCopyableOrConvertible<ValueType>();
-        if constexpr (detail::isRegisteredType<ValueType>) {
-            setArrayCopyImpl(array.begin(), array.end(), opcua::getDataType<ValueType>());
-        } else {
-            setArrayCopyConvertImpl(array.begin(), array.end());
-        }
+        static_assert(!isTemporaryArray<decltype(array)>());
+        setArrayImpl(std::data(array), std::size(array), dataType, UA_VARIANT_DATA_NODELETE);
     }
 
-    /// @overload
-    template <typename ArrayLike, typename = EnableIfNoSpan<ArrayLike>>
+    /**
+     * Copy array to variant.
+     * @param array Iterable container, for example `std::vector`, `std::list` or `Span`.
+     *              The container must implement `begin()` and `end()`.
+     */
+    template <typename ArrayLike>
     void setArrayCopy(ArrayLike&& array) {
```

Comment:
It is not a bug to use forwarding references here, but const references also bind to anything and communicate intend better.
The parameter array is not modified inside setArrayCopy and can therefore be const.
Using const references also reduces the number of template instantiations.
We get just one instantiation, instead of 3 (&&, const & and &).

Reply:
Very true. I didn't think about the multiple template instantiations 👍

Reply:
Fixed in https://github.com/open62541pp/open62541pp/pull/174/commits/699d59e73c44f5aa38e93fc86fceb30d27ef4cb3